src/smclient/smclient.hpp

```diff
@@ -0,0 +1,193 @@
+     *
+     * @return Error.
+     */
+    Error InstancesUpdateStatus(const Array<InstanceStatus>& instances) override;
+
+    /**
+     * Destroys object instance.
+     */
+    ~SMClient() = default;
+
+private:
+    using StubPtr = std::unique_ptr<SMProto::SMService::StubInterface>;
+    using StreamPtr
+        = std::unique_ptr<grpc::ClientReaderWriterInterface<SMProto::SMOutgoingMessages, SMProto::SMIncomingMessages>>;
+
+    std::unique_ptr<grpc::ClientContext> CreateClientContext();
+    StubPtr CreateStub(const std::string& url, const std::shared_ptr<grpc::ChannelCredentials>& credentials);
+
+    bool SendNodeConfigStatus(const String& version, const Error& configErr);
+
+    bool RegisterSM(const std::string& url);
+    void ConnectionLoop() noexcept;
+    void HandleIncomingMessages() noexcept;
+
+    bool ProcessGetNodeConfigStatus();
+    bool ProcessCheckNodeConfig(const SMProto::CheckNodeConfig& request);
+    bool ProcessSetNodeConfig(const SMProto::SetNodeConfig& request);
+    bool ProcessRunInstances(const SMProto::RunInstances& request);
+    bool ProcessUpdateNetworks(const SMProto::UpdateNetworks& request);
+    bool ProcessGetSystemLogRequest(const SMProto::SystemLogRequest& request);
+    bool ProcessGetInstanceLogRequest(const SMProto::InstanceLogRequest& request);
+    bool ProcessGetInstanceCrashLogRequest(const SMProto::InstanceCrashLogRequest& request);
+    bool ProcessOverrideEnvVars(const SMProto::OverrideEnvVars& request);
+    bool ProcessGetAverageMonitoring();
+    bool ProcessConnectionStatus(const SMProto::ConnectionStatus& request);
+
+    config::Config                              mConfig           = {};
+    iam::provisionmanager::ProvisionManagerItf* mProvisionManager = nullptr;
+    crypto::CertLoaderItf*                      mCertLoader       = nullptr;
+    crypto::x509::ProviderItf*                  mCryptoProvider   = nullptr;
+    iam::nodeinfoprovider::NodeInfoProviderItf* mNodeInfoProvider = nullptr;
+    sm::resourcemanager::ResourceManagerItf*    mResourceManager  = nullptr;
+    sm::networkmanager::NetworkManagerItf*      mNetworkManager   = nullptr;
+    sm::logprovider::LogProviderItf*            mLogProvider      = nullptr;
+    monitoring::ResourceMonitorItf*             mResourceMonitor  = nullptr;
+    sm::launcher::LauncherItf*                  mLauncher         = nullptr;
+
+    std::vector<std::shared_ptr<grpc::ChannelCredentials>> mCredentialList;
+    bool                                                   mCredentialListUpdated = false;
+    bool                                                   mProvisioningMode      = false;
+    NodeInfo                                               mNodeInfo;
+
+    std::unique_ptr<grpc::ClientContext> mCtx;
+    StreamPtr                            mStream;
+    StubPtr                              mStub;
+
+    std::thread             mConnectionThread;
+    std::mutex              mMutex;
+    std::atomic_bool        mStopped = true;
```

Comment:
Consider to avoid using atomic_bool as you already have mutex and `mStopped` already guarded with mutex in most the cases.